src/lib.rs

```diff
@@ -76,28 +86,372 @@ impl ULinkZenoh {
+            return Err(RpcMapperError::ProtobufError(String::from(
+                "Unable to encode UAttributes",
+            )));
+        };
+
+        // Add attachment and payload
+        let mut attachment = AttachmentBuilder::new();
+        attachment.insert("uattributes", attr.as_slice());
+        let value = Value::new(buf.into()).encoding(Encoding::WithSuffix(
+            KnownEncoding::AppCustom,
+            payload.format.to_string().into(),
+        ));
+        // TODO: Query should support .encoding
+        // TODO: Adjust the timeout
+        let getbuilder = self
+            .session
+            .get(&zenoh_key)
+            .with_value(value)
+            .with_attachment(attachment.build())
+            .target(QueryTarget::BestMatching)
+            .timeout(Duration::from_millis(1000));
+
+        // Send the query
+        let Ok(replies) = getbuilder.res().await else {
+            return Err(RpcMapperError::UnexpectedError(String::from(
+                "Error while sending Zenoh query",
+            )));
+        };
+
+        let Ok(reply) = replies.recv_async().await else {
             return Err(RpcMapperError::UnexpectedError(String::from(
-                "Wrong UAttributes",
+                "Error while receiving Zenoh reply",
             )));
+        };
+        match reply.sample {
+            Ok(sample) => {
+                let Ok(encoding) = sample.value.encoding.suffix().parse::<i32>() else {
+                    return Err(RpcMapperError::UnexpectedError(String::from(
+                        "Error while parsing Zenoh encoding",
+                    )));
+                };
+                Ok(UPayload {
+                    length: Some(0),
+                    format: encoding,
+                    data: Some(Data::Value(sample.payload.contiguous().to_vec())),
+                })
+            }
+            Err(_) => Err(RpcMapperError::UnexpectedError(String::from(
+                "Error while parsing Zenoh reply",
+            ))),
         }
+    }
+}
 
-        // TODO: Not implemented
-        Ok(payload)
+#[async_trait]
+impl RpcServer for ULinkZenoh {
+    async fn register_rpc_listener(
+        &self,
+        method: UUri,
+        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
+    ) -> Result<String, UStatus> {
+        // Do the validation
+        // TODO: Whether we should check URI is resolved or not
+        //       https://github.com/eclipse-uprotocol/uprotocol-spec/issues/42#issuecomment-1882273981
+        if UriValidator::validate(&method).is_err() {
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Invalid topic",
+            ));
+        }
+
+        // Get Zenoh key
+        let zenoh_key = ULinkZenoh::to_zenoh_key_string(&method)?;
+        // Generate listener string for users to delete
+        let mut hashmap_key = format!("{}_{:X}", zenoh_key, rand::random::<u64>());
```

Comment:
The id could be a `64bit` counter. E.g.: a static atomic.

Reply:
Fixed